test/e2e/syncer/spec_test.go

```diff
@@ -0,0 +1,254 @@
+
+				coreSharedInformerFactory := coreexternalversions.NewSharedInformerFactoryWithOptions(servers[sinkClusterName].k8sClientSet, 0)
+				nsInformer := coreSharedInformerFactory.Core().V1().Namespaces()
+				nsExpecter := framework.NewExpecter(nsInformer.Informer())
+				sinkNsLister := nsInformer.Lister()
+				coreSharedInformerFactory.Start(ctx.Done())
+
+				wildwestSharedInformerFactory := wildwestexternalversions.NewSharedInformerFactoryWithOptions(servers[sinkClusterName].wildwestClientSet, 0)
+				cowboyInformer := wildwestSharedInformerFactory.Wildwest().V1alpha1().Cowboys()
+				cowboyExpecter := framework.NewExpecter(cowboyInformer.Informer())
+				sinkCowboyLister := cowboyInformer.Lister().Cowboys("")
+				wildwestSharedInformerFactory.Start(ctx.Done())
+
+				ns, err := sourceNsClient.Create(ctx, &corev1.Namespace{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: testNamespace,
+					},
+				}, metav1.CreateOptions{})
+				if err != nil {
+					t.Errorf("failed to create namespace: %v", err)
+					return
+				}
+
+				cowboy, err := sourceCowboyClient.Create(ctx, &wildwestv1alpha1.Cowboy{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "timothy",
+					},
+					Spec: wildwestv1alpha1.CowboySpec{Intent: "yeehaw"},
+				}, metav1.CreateOptions{})
+				if err != nil {
+					t.Errorf("failed to create cowboy: %v", err)
+					return
+				}
+
+				if !cache.WaitForNamedCacheSync(t.Name(), ctx.Done(), nsInformer.Informer().HasSynced) {
+					t.Error("failed to wait for namespace cache to sync")
+					return
+				}
+
+				if !cache.WaitForNamedCacheSync(t.Name(), ctx.Done(), cowboyInformer.Informer().HasSynced) {
+					t.Error("failed to wait for cowboy cache to sync")
+					return
+				}
+
+				defer servers[sourceClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sourceNsClient.Get(ctx, ns.Name, metav1.GetOptions{})
+				})
+				defer servers[sourceClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sourceCowboyClient.Get(ctx, cowboy.Name, metav1.GetOptions{})
+				})
+				defer servers[sinkClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sinkNsClient.Get(ctx, fmt.Sprintf("kcp--admin--%s", ns.Name), metav1.GetOptions{})
+				})
+				defer servers[sinkClusterName].Artifact(t, func() (runtime.Object, error) {
+					return sinkWildwestClient.Cowboys(fmt.Sprintf("kcp--admin--%s", cowboy.Name)).Get(ctx, cowboy.GetName(), metav1.GetOptions{})
+				})
+
+				if err := nsExpecter.ExpectBefore(ctx, func(ctx context.Context) (bool, error) {
+					sinkNsList, err := sinkNsLister.List(labels.Everything())
+					if err != nil {
+						return !apierrors.IsNotFound(err), err
+					}
+					var names []string
+					for _, ns := range sinkNsList {
+						names = append(names, ns.Name)
+						if ns.Name == expectedSinkNsName {
+							return true, nil
+						}
+					}
+					return false, fmt.Errorf("No namespace in sink cluster: %v", names)
```

Comment:
I was initially wondering why it made sense to use a list instead of a get, and the answer appears to be so that this error message can contain a list of all namespaces in the target cluster. Why is that useful in this case vs indicating something like 'namespace %s not present'?

Reply:
The original reason is that I copy/pasted from a different e2e test, and that test was listing.  But I will say that having the list of existing namespaces did help me debug the test.  I should probably at least add the expected name to the log message.

Reply:
I can see the list of existing namespaces helping if the intent is to validate more than just the creation of the namespace. But if the goal is validating the metadata transformation, maybe that could tested with less cost in a unit test?

Reply:
If you want to have the list of namespaces after the test you can also certainly add them as an artifact